File: controllers/authController.js
import User from "../models/User.js";
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";

/* ================= REGISTER ================= */
export const register = async (req, res) => {
  const { name, email, password } = req.body;

  // check user exists
  const existing = await User.findOne({ email });
  if (existing) {
    return res.status(400).json({ message: "User already exists" });
  }

  // password encrypt
  const hashedPassword = await bcrypt.hash(password, 10);

  // save user
  await User.create({
    name,
    email,
    password: hashedPassword
  });

  res.json({ message: "User registered successfully" });
};

/* ================= LOGIN ================= */
export const login = async (req, res) => {
  const { email, password } = req.body;

  const user = await User.findOne({ email });
  if (!user) {
    return res.status(404).json({ message: "User not found" });
  }

  // password match
  const isMatch = await bcrypt.compare(password, user.password);
  if (!isMatch) {
    return res.status(401).json({ message: "Wrong password" });
  }

  // JWT token
  const token = jwt.sign(
    { id: user._id },
    process.env.JWT_SECRET,
    { expiresIn: "7d" }
  );

  res.json({ token, user });
};

File: controllers/auth_controller.py
import os
import bcrypt
from datetime import datetime, timedelta
from fastapi import HTTPException, status
from jose import jwt
from pydantic import BaseModel
from models.user import User
from dotenv import load_dotenv

load_dotenv()


# ================= REQUEST SCHEMAS =================

class RegisterRequest(BaseModel):
    name: str
    email: str
    password: str


class LoginRequest(BaseModel):
    email: str
    password: str


# ================= REGISTER =================

async def register(data: RegisterRequest):
    existing = await User.find_one(User.email == data.email)
    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="User already exists"
        )

    hashed = bcrypt.hashpw(data.password.encode("utf-8"), bcrypt.gensalt())
    hashed_password = hashed.decode("utf-8")

    user = User(name=data.name, email=data.email, password=hashed_password)
    await user.insert()

    return {"message": "User registered successfully"}


# ================= LOGIN =================

async def login(data: LoginRequest):
    user = await User.find_one(User.email == data.email)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )

    is_match = bcrypt.checkpw(
        data.password.encode("utf-8"),
        user.password.encode("utf-8")
    )
    if not is_match:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Wrong password"
        )

    secret = os.getenv("JWT_SECRET", "mysecretkey123")
    token = jwt.encode(
        {"id": str(user.id), "exp": datetime.utcnow() + timedelta(days=7)},
        secret,
        algorithm="HS256"
    )

    user_data = user.dict()
    user_data["_id"] = str(user.id)
    user_data.pop("password", None)

    return {"token": token, "user": user_data}
